### PR TITLE
Allow configuration of PushConfiguration on server side

### PR DESCRIPTION
**Pull request by mcollovati**
Allow configuration of PushConfiguration parameters and
the possibility to specify a custom PushConnection
implementations on server side (#3752).

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/flow/3819)

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java [hunk 1]
```diff
@@ -161,18 +164,50 @@
...
+    }
+
+    /**
+     * Interface to be implemented by {@link PushConfiguration} classes
+     * that wishes to delegate the creation of {@link PushConnection}
+     * to an external component.
+     *
+     * @since
+     */
+    interface WithConnectionFactory {
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java]: Not sure it makes sense to mark this interface as @FunctionalInterface

**Comment by Artur-** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java, line 174 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9JvEuk2Idd4A_aURA8:-L9JvEuk2Idd4A_aURA9:b-c2la3m) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java#L174)):*
> ```Java
>      * @param pushConnectionFactory The push connection factory
>      */
>     default void applyConnectionFactoryIfPossible(SerializableFunction<UI, PushConnection> pushConnectionFactory) {
> ```

Why is this needed? Can't `setPushConnectionFactory` be here instead?

---

*[flow-server/src/main/java/com/vaadin/flow/component/page/Push.java, line 68 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9Kgyh--eeJJnefBjNw:-L9Kgyh--eeJJnefBjNx:b-u9rrbl) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/component/page/Push.java#L68)):*
> ```Java
>      * @since
>      */
>     Class<? extends PushConfigurator> configurator() default DefaultPushConfigurator.class;
> ```

Somehow looks like a strange place to define this. Would you use different configurators for different views? Looks more like a global setting which you would set once

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by mcollovati** [comment]
*[flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java, line 174 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9JvEuk2Idd4A_aURA8:-L9L8MtU0VJCn0uzG2LC:bgn5sof) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/component/PushConfiguration.java#L174)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Why is this needed? Can't `setPushConnectionFactory` be here instead?
</blockquote></details>

`PushConfiguration` is implemented also by`PushConfigurationMap` and my thought was that it doesn't make sense to have a no-op`setPushConnectionFactory` there. So this method is simply a helper for `PushConfigurator` implementations to avoid code duplication.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by mcollovati** [comment]
*[flow-server/src/main/java/com/vaadin/flow/component/page/Push.java, line 68 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9Kgyh--eeJJnefBjNw:-L9LBbA74Ew9dRAvWpLy:bijqa03) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/component/page/Push.java#L68)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Somehow looks like a strange place to define this. Would you use different configurators for different views? Looks more like a global setting which you would set once
</blockquote></details>

It seemed to me that `@Push` annotation is used only in `BootstrapHandler` so my thought was that a configurator is needed only where push settings are explicitly declared.
Is `PushConfiguration` initialized or reset in other places?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by mcollovati** [comment]
*[flow-server/src/main/java/com/vaadin/flow/server/communication/DefaultPushConfigurator.java, line 36 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9oOca782yfbPc-LkFF:-L9oaIDJ2beq6FhRi9I2:b-pa8eoz) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/server/communication/DefaultPushConfigurator.java#L36)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

With this design, there will always be a `@Push` annotation and always a configuration instance that implements `PushConfiguration`. It feels weird that the way values are copied from the annotation to the configuration would be different with different configurators.

From this would also follow a possibility for further simplification: `PushConfigurator` wouldn't be responsible for anything than defining how a `PushConnection` is created.
</blockquote></details>

So `PushConfigurator` could be replaced by a `PushConnectionFactory : Function<UI, PushConnection>` and the implementation could be loaded via `ServiceLoader` directly in `PushConfigurationImpl` leaving `BootstrapHandler` as is.
If no provider is found AtmosphereConnection will be used and if there are more than one provider an exception will be thrown.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by Legioth** [comment]
*[flow-server/src/main/java/com/vaadin/flow/server/communication/DefaultPushConfigurator.java, line 36 at r2](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9oOca782yfbPc-LkFF:-L9opFc82DigLQlTnDzz:b-d5jd5b) ([raw file](https://github.com/vaadin/flow/blob/d2ebdb56a6bc14226823a931c76ddd2b79d05610/flow-server/src/main/java/com/vaadin/flow/server/communication/DefaultPushConfigurator.java#L36)):*
<details><summary><i>Previously, mcollovati (Marco Collovati) wrote…</i></summary><blockquote>

So `PushConfigurator` could be replaced by a `PushConnectionFactory : Function<UI, PushConnection>` and the implementation could be loaded via `ServiceLoader` directly in `PushConfigurationImpl` leaving `BootstrapHandler` as is.
If no provider is found AtmosphereConnection will be used and if there are more than one provider an exception will be thrown.
</blockquote></details>

Good point that both simplifications could be combined.

The main drawback with using `ServiceLoader` is that it's not very discoverable and has the potential for causing unexpected conflicts. The benefit is that it would be really easy to use and "hide" this quite special case from the API used by regular users.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by Legioth** [comment]
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---

*[flow-server/src/main/java/com/vaadin/flow/server/communication/PushConnectionFactory.java, line 28 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9skJ6a2d-9vBNkakH9:-L9skJ6a2d-9vBNkakHA:b7usngf) ([raw file](https://github.com/vaadin/flow/blob/29464313232b0c21491e8d0a63437d46584bdaae/flow-server/src/main/java/com/vaadin/flow/server/communication/PushConnectionFactory.java#L28)):*
> ```Java
>  * Produces a {@link PushConnection} for the provided {@link UI}.
>  *
>  * Factory instances are by default discovered and instantiated using
> ```

Either remove "by default" or make it possible to override the default by making the `setupPushConnectionFactory` method `protected`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by mcollovati** [comment]
*[flow-server/src/main/java/com/vaadin/flow/server/communication/PushConnectionFactory.java, line 28 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3819#-L9skJ6a2d-9vBNkakH9:-L9sr0xL0tlRbQW6i1Bn:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/29464313232b0c21491e8d0a63437d46584bdaae/flow-server/src/main/java/com/vaadin/flow/server/communication/PushConnectionFactory.java#L28)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Either remove "by default" or make it possible to override the default by making the `setupPushConnectionFactory` method `protected`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819)*

**Comment by SomeoneToIgnore** [comment]
:lgtm:

---

Reviewed 4 of 11 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3819#-:-LADR7Jw9tBrFZFLSx2d:bnfp4nl)*

**Pull request merged**